uefi/src/fs/file_system.rs

```diff
@@ -0,0 +1,312 @@
+//! Module for [FileSystem].
+
+use super::*;
+use super::dir_entry_iter::UefiDirectoryIter;
+use super::normalized_path::{NormalizedPath, PathError};
+use super::path::Path;
+use alloc::boxed::Box;
+use alloc::string::{FromUtf8Error, String, ToString};
+use alloc::vec::Vec;
+use core::fmt;
+use core::fmt::{Debug, Formatter};
+use core::ops::Deref;
+use derive_more::Display;
+use uefi::proto::media::file::{FileAttribute, FileInfo, FileType};
+use uefi::table::boot::ScopedProtocol;
+use alloc::format;
+use log::info;
+use uefi::fs::path::Component;
+use crate::fs::path::Components;
+
+/// All errors that can happen when working with the [`FileSystem`].
+#[derive(Debug, Clone, Display, PartialEq, Eq)]
+pub enum FileSystemError {
+    /// Can't open the root directory of the underlying volume.
+    CantOpenVolume,
+    /// The path is invalid because of the underlying [`PathError`].
+    IllegalPath(PathError),
+    /// The file or directory was not found in the underlying volume.
+    FileNotFound(String),
+    /// The path is existend but does not correspond to a directory when a directory was expected.
+    NotADirectory(String),
+    /// The path is existend but does not correspond to a file when a file was expected.
+    NotAFile(String),
+    /// Can't delete the file.
+    CantDeleteFile(String),
+    /// Can't delete the directory.
+    CantDeleteDirectory(String),
+    /// Error writing bytes.
+    WriteFailure,
+    /// Error flushing file.
+    FlushFailure,
+    /// Error reading file.
+    ReadFailure,
+    /// Can't parse file content as UTF-8.
+    Utf8Error(FromUtf8Error),
+    /// Could not open the given path.
+    OpenError(String),
+}
+
+#[cfg(feature = "unstable")]
+impl core::error::Error for FileSystemError {}
+
+impl From<PathError> for FileSystemError {
+    fn from(err: PathError) -> Self {
+        Self::IllegalPath(err)
+    }
+}
+
+/// Return type for public [`FileSystem`] operations.
+pub type FileSystemResult<T> = Result<T, FileSystemError>;
+
+/// High-level file-system abstraction for UEFI volumes with an API that is
+/// close to `std::fs`. It acts as convenient accessor around the
+/// [`SimpleFileSystemProtocol`].
+pub struct FileSystem<'boot_services> {
```

Comment:
this lifetime is technically wrong. It only lifes as long "as it lives". Should be the same the underlying protocol, hence, `'a`.